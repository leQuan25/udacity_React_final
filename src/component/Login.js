import { useState } from "react";
import { handleInitialData } from "../actions/shared"
import { useEffect } from "react";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";
import { useNavigate } from "react-router-dom";
import loginImage from "../images/Login.png"

const Login = (props) => {
    const navigate = useNavigate();
    const [userName, setUserName] = useState("");
    const [passWord, setPassWord] = useState("");
  
    useEffect(() => {
        props.dispatch(handleInitialData());
      }, []);


    const handleChangeUserName = (e) => {
        const text = e.target.value;
        setUserName(text);
      };

    const handleChangePassWord = (e) => {
        const text = e.target.value;
        setPassWord(text);
      };
   
      function handleLogin(userName, password, userList){
        for (const key in userList.users) {
            if (userList.users[key].id === userName && userList.users[key].password === password) {
                return true;
            }
        }
        return false;
    }

      const handleSubmit = (e) => {
        e.preventDefault();        
        if (handleLogin(userName, passWord, props.userList)) {
            props.dispatch(setAuthedUser(userName));
            navigate("/home");
        } else {
            alert("Wrong user or password");
        }
      };

    return (
        <div>
            <h3>Employee Polls</h3>
            <img src={loginImage} />
            <h3>Login</h3>

        <form className="login" onSubmit={handleSubmit}>
            <label htmlFor="username">Username:</label>
            <input 
            type="text" 
            id="username"
            data-testid="username"
            name="username" 
            placeholder="User" 
            value={userName}
            onChange={handleChangeUserName} />
            <br></br>
            
            <label htmlFor="password">Password:</label>
            <input 
            type="password" 
            id="password" 
            data-testid="password"
            name="password" 
            placeholder="Password"
            value={passWord}
            onChange={handleChangePassWord}
            />
            <br></br>
            
            <button data-testid="submitButton" className="btn" type="submit" disabled={userName === "" && passWord === ""}>
            Submit
            </button>
        </form>  
        </div>
    );
};


function mapStateToProps(state) {
    const { users } = state;
    return { userList: users }
} 

export default connect(mapStateToProps)(Login);