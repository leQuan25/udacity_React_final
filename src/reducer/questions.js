import { RECEIVE_QUESTIONS, TOGGLE_QUESTIONS, ADD_QUESTIONS } from "../actions/questions";

export default function questions(state = {}, action) {
  switch (action.type) {
    case RECEIVE_QUESTIONS:
    return {
        ...state,
        questions: action.questions,
      };

      case TOGGLE_QUESTIONS:
      return {
        ...state,
        [action.qid]: {
            ...state.questions[action.qid],
            optionOne:
                action.answer === 'optionOne' ? state.questions[action.qid].optionOne.votes.concat([action.authedUser]) : state.questions[action.qid].votes,
            optionTwo:
                action.answer === 'optionTwo' ? state.questions[action.qid].optionTwo.votes.concat([action.authedUser]) : state.questions[action.qid].votes,
        },
      };

     case ADD_QUESTIONS:
        const { questions } = action;
        return {
            ...state,
            ...questions,
        };

    default:
      return state;
  }
}
