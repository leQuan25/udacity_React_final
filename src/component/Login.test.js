import {render, fireEvent} from '@testing-library/react';
import * as React from 'react';
import Login from './Login';
import reducer from '../reducer';
import middleware from '../middleware';
import {legacy_createStore as createStore} from 'redux';
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";

describe('Login', () => {
    it('will match snapshot', () => {
        const store = createStore(reducer, middleware);
        var component = render(<Provider store={store}><Router><Login/></Router></Provider>);
        expect(component).toMatchSnapshot();
    });
});

describe('Login page display items check', () => {
 it('check disable and enable items', () => {
    const store = createStore(reducer, middleware);
        var component = render(<Provider store={store}><Router><Login/></Router></Provider>);
        expect(component.getByTestId('submitButton')).toHaveAttribute('disabled');
        var userName = component.getByTestId('username');
        var password = component.getByTestId('password');
        fireEvent.change(userName, {target: { value: 'userName'}});
        fireEvent.change(password, {target: { value: 'password'}});
        expect(component.getByTestId('submitButton')).toBeEnabled();
 });  
});

