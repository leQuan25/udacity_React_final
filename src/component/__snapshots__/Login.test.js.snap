// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h3>
          Employee Polls
        </h3>
        <img
          src="Login.png"
        />
        <h3>
          Login
        </h3>
        <form
          class="login"
        >
          <label
            for="username"
          >
            Username:
          </label>
          <input
            data-testid="username"
            id="username"
            name="username"
            placeholder="User"
            type="text"
            value=""
          />
          <br />
          <label
            for="password"
          >
            Password:
          </label>
          <input
            data-testid="password"
            id="password"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <br />
          <button
            class="btn"
            data-testid="submitButton"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h3>
        Employee Polls
      </h3>
      <img
        src="Login.png"
      />
      <h3>
        Login
      </h3>
      <form
        class="login"
      >
        <label
          for="username"
        >
          Username:
        </label>
        <input
          data-testid="username"
          id="username"
          name="username"
          placeholder="User"
          type="text"
          value=""
        />
        <br />
        <label
          for="password"
        >
          Password:
        </label>
        <input
          data-testid="password"
          id="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <br />
        <button
          class="btn"
          data-testid="submitButton"
          disabled=""
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
